tensorforce/models/ppo_model.py

```diff
@@ -171,18 +174,45 @@ def update(self, batch):
                 fetches.update(self.distribution_tensors)
                 prev_distribution_tensors = self.session.run(fetches=fetches, feed_dict=feed_dict)
                 prev_distribution_tensors.pop('optimize')
-
-            elif i == self.updates - 1:  # Last update, fetch return and diagnostics values
-                fetches = (self.optimize, self.loss, self.loss_per_instance, self.kl_divergence, self.entropy)
-                prev_distribution_tensors = {placeholder: tensor for name, placeholders in self.prev_distribution_tensors.items() for placeholder, tensor in zip(placeholders, prev_distribution_tensors[name])}
-                feed_dict.update(prev_distribution_tensors)
-                loss, loss_per_instance, kl_divergence, entropy = self.session.run(fetches=fetches, feed_dict=feed_dict)[1:]
-
             else:  # Otherwise just optimize
                 self.session.run(fetches=self.optimize, feed_dict=feed_dict)
 
+        # For the last epoch, fetch return and diagnostics values for each instance
```

Comment:
We need to return the loss for each instance. Prior to this the agent was only returning the values from one minibatch. You can't do this with random sampling (although random sampling is nice idea with PPO) so for the last epoch I go through sequential mini batches of the whole batch.

Suggestion to refactor this code are welcome.